### PR TITLE
Subscribe to docker stats and place some in the Container topology.

### PR DESCRIPTION
**Pull request by tomwilkie**
Work towards #114

### CODE REVIEW

File: probe/main.go [hunk 1]
```diff
@@ -144,8 +144,18 @@ func hostTopology(hostID, hostName string) report.Topology {
...
+	for {
+		switch <-sigs {
+		case syscall.SIGINT:
+			log.Printf("=== received SIGINT ===\n*** exiting\n")
+			return
+		case syscall.SIGQUIT:
+			stacklen := runtime.Stack(buf, true)
+			log.Printf("=== received SIGQUIT ===\n*** goroutine dump...\n%s\n*** end\n", buf[:stacklen])
+		}
+	}
```
Comment (peterbourgon) [inline probe/main.go]: If you simply don't trap SIGQUIT, you get this behavior by default in the runtime.

**Comment by tomwilkie** [comment]
Note this introduces a startup lag (as it connects all the stats go routines on the main thread).  This should be moved off to a background thread (and parallelised?)

File: probe/tag/docker_tagger.go [hunk 2]
```diff
@@ -157,15 +179,15 @@ func (t *DockerTagger) update() bool {
...
 
-func (t *DockerTagger) updateContainers(client dockerClient) error {
-	apiContainers, err := client.ListContainers(docker.ListContainersOptions{All: true})
+func (t *DockerTagger) updateContainers() error {
+	apiContainers, err := t.client.ListContainers(docker.ListContainersOptions{All: true})
 	if err != nil {
 		return err
 	}
 
 	containers := []*docker.Container{}
```
Comment (peterbourgon) [inline probe/tag/docker_tagger.go]: t.update never returns unless there's an error during set up, or the user issues a quit. But you call it repeatedly in loop, and always on a ticker. Why?

Comment (tomwilkie) [inline probe/tag/docker_tagger.go]: To make it retry if it fails; if a connection is dropped for instance, or Docker is restarted.

File: probe/main.go (hunk 1 above)
Comment (tomwilkie) [inline probe/main.go]: D'oh.  Weave has this everywhere, so I assumed not.  Might be worth filing a bug with weave too.

**Comment by peterbourgon** [comment]
There's a lot of manual mutex locking going on. Since you already have a goroutine handling Docker events, why not thread all user requests through it, too? Then you can eliminate all the locks.

**Comment by tomwilkie** [comment]
Yes the locking makes me nervous; going to an actor model seems appropriate here.

I was always considering having our own Container struct, and moving all the stats & updating stuff into that.  Then docker mapper is responsible for creating / removing containers in response to events.

We should probably also break out the mapper & topology logic to be a client of this type.

File: probe/tag/docker_tagger.go [hunk 3]
```diff
@@ -249,6 +295,79 @@ func (t *DockerTagger) updatePIDTree() error {
...
+	}
+
+	conn.Close()
+	delete(t.statsConns, containerID)
+	delete(t.latestStats, containerID)
+	return
+}
+
 // Containers returns the Containers the DockerTagger knows about.
 func (t *DockerTagger) Containers() []*docker.Container {
```
Comment (peterbourgon) [inline probe/tag/docker_tagger.go]: I think this is dead code?

File: probe/tag/docker_tagger.go [hunk 4]
```diff
@@ -336,8 +455,42 @@ func (t *DockerTagger) ContainerTopology(scope string) report.Topology {
...
 		result.NodeMetadatas[nodeID] = nmd
 	}
 	return result
 }
+
+// called whilst holding t.RLock()
+func (t *DockerTagger) getStats(containerID string) map[string]string {
+	stats, ok := t.latestStats[containerID]
+	if !ok {
+		return nil
```
Comment (peterbourgon) [inline probe/tag/docker_tagger.go]:
```go
return map[string]string{}
```

is safer :)

File: probe/tag/docker_tagger.go (hunk 3 above)
Comment (tomwilkie) [inline probe/tag/docker_tagger.go]: Used by tracer.

**Comment by peterbourgon** [comment]
I've been reading this for awhile and am having trouble building and keeping a mental model of what's happening. There's a lot of action at a distance, mutating the DockerTagger struct in-place. The preponderance of mutex-locking is an artifact of that. I think it would benefit from a single loop() method handling user requests, events from the Docker daemon, and events from individual /container/foo/stats streams.

**Comment by tomwilkie** [comment]
@peterbourgon I didn't go full actor on this, but I've rationalised the flow and locking.

Let me know what you think.  There is a lot of potential for refactoring here, but I'd prefer to save anything major for another PR.

File: probe/tag/docker_tagger.go [hunk 5]
```diff
@@ -249,13 +217,52 @@ func (t *DockerTagger) updatePIDTree() error {
...
 }
 
+func (t *DockerTagger) addContainer(containerID string) error {
+	container, err := t.client.InspectContainer(containerID)
+	if err != nil {
+		// Don't spam the logs if the container was short lived
+		if _, ok := err.(*docker.NoSuchContainer); !ok {
+			return err
+		}
+		return nil
```
Comment (peterbourgon) [inline probe/tag/docker_tagger.go]: Hmm, should this be inverted?

```go
if _, ok := err.(*docker.NoSuchContainer); ok {
    return nil
}
return err
```

**Comment by peterbourgon** [comment]
Much, much easier to understand. Thanks a bunch. LGTM.

**Comment by tomwilkie** [comment]
Going to hold off on this until post 0.3

**Pull request closed without merging**